File: 22-08/19/871.cpp
#include <vector>
#include <iostream>
using namespace std;


class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int> >& stations) {
        int n = stations.size();
        vector<vector<int> > dp(n + 1, vector<int>(target + 1, 0));

        for (int i = 0; i <= n; ++i) {
            dp[i][0] = 0;
        }

        for (int i = 1; i <=n; i++){
            for (int j=1; j <=i; j++){
                dp[i][j] = dp[i-1][j];
                if (j >= stations[i-1][0]){
                    dp[i][j] = max(dp[i][j], dp[i-1][j-1] + stations[i-1][1]);
                }
            }
        }


        for (int j = 0; j < n+1; j++){
            if (dp[n][j] >= target){
                return j;
            }
        }
        return -1;
    }
};

// create a test case to test the solution
// target = 1, startFuel = 1, stations = [[0,1]]
// output = 1
// target = 100, startFuel = 1, stations = [[0,1],[10,100]]
// output = -1
// target = 100, startFuel = 10, stations = [[0,1],[10,100]]
// output = 1
// target = 100, startFuel = 10, stations = [[0,1],[10,100],[20,200],[30,300],[40,400],[50,500],[60,600],[70,700],[80,800],[90,900],[100,1000]]
// output = 7
// main function to test the solution
int main() {
    Solution s;
    int target = 100;
    int startFuel = 10;
    vector<vector<int> > stations {{0,1},{10,100},{20,200},{30,300},{40,400},{50,500},{60,600},{70,700},{80,800},{90,900},{100,1000}};
    cout << s.minRefuelStops(target, startFuel, stations) << endl;
    return 0;
}

File: 22-08/19/871.py
from typing import List, Tuple


class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> Tuple[int, List[int]]:
        # Use dp[i][j] to represent the farthest distance it can get at station i and with j stops
        n = len(stations)
        dp = [[0] * (n + 1) for _ in range(n + 1)]
        for i in range(n+1):
            dp[i][0] = startFuel

        for i in range(1, n + 1):
            for j in range(1, i + 1):
                dp[i][j] = dp[i - 1][j]
                if dp[i - 1][j - 1] >= stations[i - 1][0]:
                    dp[i][j] = max(dp[i][j], dp[i - 1][j - 1] + stations[i - 1][1])
        print(dp)

        for j in range(i + 1):
            if dp[i][j] >= target:
                return j

        return -1

        # Trace back the path of the solution from dp[n][k]
        # path = []
        # i = n
        # j = k
        # while i > 0:
        #     if dp[i][j] == dp[i - 1][j]:
        #         i -= 1
        #     else:
        #         path.append(i)
        #         i -= 1
        #         j -= 1
        # return k, path[::-1]


if __name__ == "__main__":
    # target = 100, startFuel = 10, stations = [[10, 60], [20, 30], [30, 30], [60, 40]]
    # In order to arrive at xn, either refill at the previous station or not
    # Use dp[i][j] to represent the farthest distance it can get at station i and with j stops
    # answer find j such that dp[n][j] >= target
    # transition function dp[i][j] = max(dp[i-1][j-1] + fuel(i), dp[i-1][j])
    # order i from 0 to n; j from 0 to n

    target = 100
    startFuel = 10
    stations = [[10, 60], [20, 30], [30, 30], [60, 40]]
    print(Solution().minRefuelStops(target, startFuel, stations))
